File: software/old_python/src/sign_hardware.py
from machine import Pin, SPI, SoftSPI, Timer
import time

# pin mappings
_EN    = const(21)
_CDATA = const(22)
_CCLK  = const(23)
_BS_IN = const(34)
_C1    = const(12)
_C2    = const(13)
_C3    = const(14)
_C4    = const(15)
_C5    = const(16)
_C6    = const(17)
_C7    = const(18)

_spi = SPI(1, baudrate=3000000, polarity=1, phase=0, bits=8, firstbit=SPI.LSB, sck=Pin(_CCLK), mosi=Pin(_CDATA))
_en = Pin(_EN, Pin.OUT, value=0)
_rows = [
    Pin(_C1, Pin.OUT, value=0),
    Pin(_C2, Pin.OUT, value=0),
    Pin(_C3, Pin.OUT, value=0),
    Pin(_C4, Pin.OUT, value=0),
    Pin(_C5, Pin.OUT, value=0),
    Pin(_C6, Pin.OUT, value=0),
    Pin(_C7, Pin.OUT, value=0)
]

def enable_output():
    global _en
    _en.on()

def disable_output():
    global _en
    _en.off()

@micropython.native
def row_on(num):
    global _rows
    _rows[num].on()

@micropython.native
def row_off(num):
    global _rows
    _rows[num].off()

def all_rows_off():
    global _rows
    for row in _rows:
        row.off()

@micropython.native
def shift_row(rownum, memory):
    global _spi
    _spi.write(memory[rownum])

File: software/firmware/main/sign_hardware.cpp
#include <Arduino.h>
#include <SPI.h>
#include "sign_memory.h"
#include "sign_hardware.h"

// pin mappings
#define _EN      21
#define _CDATA   22
#define _CCLK    23
// SS for SPI isnt actually used, we just pick an unused pin and don't even ever configure it
#define _BS_SS   27
// same story for this BS MISO pin
#define _BS_MISO 27
// #define _BS_IN 34
#define _C1      12
#define _C2      13
#define _C3      14
#define _C4      15
#define _C5      16
#define _C6      17
#define _C7      18

#define SPI_FREQ 3250000
uint8_t row_pins[] = {
  _C1, _C2, _C3, _C4, _C5, _C6, _C7
};

SPIClass spi;
SPISettings spiSettings;

void init_sign_hardware(){
  pinMode(_EN, OUTPUT);
  digitalWrite(_EN, LOW);
  pinMode(_CDATA, OUTPUT);
  pinMode(_CCLK, OUTPUT);
  pinMode(_C1, OUTPUT);
  pinMode(_C2, OUTPUT);
  pinMode(_C3, OUTPUT);
  pinMode(_C4, OUTPUT);
  pinMode(_C5, OUTPUT);
  pinMode(_C6, OUTPUT);
  pinMode(_C7, OUTPUT);
  spiSettings = SPISettings(SPI_FREQ, LSBFIRST, SPI_MODE0);
  spi = SPIClass(HSPI);
  spi.begin(_CCLK, _BS_MISO, _CDATA, _BS_SS);
  spi.setFrequency(SPI_FREQ);
  spi.setBitOrder(LSBFIRST);
}

void enable_output(){
  digitalWrite(_EN, HIGH);
}

void disable_output(){
  digitalWrite(_EN, LOW);
}

void row_on(uint8_t rownum){
  digitalWrite(row_pins[rownum], HIGH);
}

void row_off(uint8_t rownum){
  digitalWrite(row_pins[rownum], LOW);
}

void all_rows_off(){
  row_off(_C1);
  row_off(_C2);
  row_off(_C3);
  row_off(_C4);
  row_off(_C5);
  row_off(_C6);
  row_off(_C7);
}

void shift_row(SIGN_ROW row){
  spi.writeBytes(row, BYTES_PER_ROW);
}
